src/rt/sched/threadsync.h

```diff
@@ -2,32 +2,143 @@
+   *  - Unlocked: No thread is currently holding the lock.
+   *  - Locked: A thread is holding the lock, and is processing its
+   *    pause/unpause request
+   *  - LockedUnpauseNeeded: A thread is holding the lock, and another thread
+   *    has attempted to lock for unpause. Rather than waiting for the lock to
+   *    become available, the thread currently holding the lock takes
+   *    responsibility for unpausing the other threads.
+   */
+  class SchedulerLock
+  {
+    enum State
+    {
+      Unlocked,
+      Locked,
+      LockedUnpauseNeeded
+    };
+
+    std::atomic<State> state{Unlocked};
+
+  public:
+    /**
+     * Acquires the lock. Spins waiting for it to be available.
+     */
+    void lock()
+    {
+      Systematic::cout() << "Locking Scheduler." << Systematic::endl;
+      auto u = Unlocked;
+      while (!state.compare_exchange_strong(u, Locked))
+      {
+        // Have to reset u, as compare exchange gives it the current value.
+        u = Unlocked;
+        while (state.load(std::memory_order_acquire) != Unlocked)
+        {
+          snmalloc::Aal::pause();
+        }
+      }
+      Systematic::cout() << "Locking Scheduler done" << Systematic::endl;
+    }
+
+    /**
+     * Attempts to release the lock.  If the lock has received an unpause
+     * request, then will return false, and continues to hold the lock.
+     * Otherwise, will return true, and the lock is actually released.
+     */
+    bool unlock()
+    {
+      assert(state.load(std::memory_order_relaxed) != Unlocked);
+      auto l = Locked;
+      return state.compare_exchange_strong(l, Unlocked);
+    }
+
+    /**
+     * Releases the lock ignoring any unpause requests.
+     */
+    void unlock_unpause()
+    {
+      assert(state.load(std::memory_order_relaxed) == LockedUnpauseNeeded);
+      state.store(Unlocked);
+    }
+
+    /**
+     * Attempt to acquire the lock to unpause threads.
+     * Returns true, if it acquires the lock, and false if it
+     * notified the thread currently holding the thread to wake up waiters.
+     */
+    bool lock_for_unpause()
+    {
+      return state.exchange(LockedUnpauseNeeded) == Unlocked;
+    }
+  };
+
+  struct LocalSync
+  {
+    pal::Semaphore sem;
+    LocalSync* next{nullptr};
+  };
+
   template<class T>
   class ThreadSync
   {
-    std::mutex m;
-    std::condition_variable cv;
+    SchedulerLock lock;
+    LocalSync* waiters = nullptr;
+
+    void unlock()
+    {
+      Systematic::cout() << "Unlock Scheduler lock" << Systematic::endl;
+
+      // Releasing the lock can pickup an unpause request
+      if (!lock.unlock())
+      {
+        Systematic::cout() << "Pending unpause" << Systematic::endl;
+
+        auto* curr = waiters;
+        waiters = nullptr;
+        // Subsequent unpause requests can be ignored as there are no more
+        // waiters as we have held the lock continuously.
+        lock.unlock_unpause();
+        // Don't need to hold the lock to wake up the waiters.

```

Comment:
If I read this correctly, it is to ensure that the state is updated before any waiters resume?

Reply:
Yeah, it is to ensure atomicity of the update with respect to waiters.